lib/pacemaker/pcmk_sched_colocation.c

```diff
@@ -787,24 +787,37 @@ pcmk__unpack_colocation(xmlNode *xml_obj, pe_working_set_t *data_set)
     }
 }
 
+/*!
+ * \internal
+ * \brief Make actions of a given type unrunnable for a given resource
+ *
+ * \param[in,out] rsc     Resource whose actions should be blocked
+ * \param[in]     task    Name of action to block
+ * \param[in]     reason  Unrunnable start action causing the block
+ */
 static void
-mark_start_blocked(pe_resource_t *rsc, pe_resource_t *reason,
-                   pe_working_set_t *data_set)
+mark_action_blocked(pe_resource_t *rsc, const char *task,
+                    const pe_resource_t *reason)
 {
     char *reason_text = crm_strdup_printf("colocation with %s", reason->id);
 
     for (GList *gIter = rsc->actions; gIter != NULL; gIter = gIter->next) {
         pe_action_t *action = (pe_action_t *) gIter->data;
 
         if (pcmk_is_set(action->flags, pe_action_runnable)
-            && pcmk__str_eq(action->task, RSC_START, pcmk__str_casei)) {
+            && pcmk__str_eq(action->task, task, pcmk__str_casei)) {
 
             pe__clear_action_flags(action, pe_action_runnable);
             pe_action_set_reason(action, reason_text, false);
-            pcmk__block_colocated_starts(action, data_set);
-            pcmk__update_action_for_orderings(action, data_set);
+            pcmk__block_colocated_starts(action, rsc->cluster);
+            pcmk__update_action_for_orderings(action, rsc->cluster);
         }
     }
+
+    // If parent resource can't perform an action, neither can any children
+    for (GList *iter = rsc->children; iter != NULL; iter = iter->next) {
+        mark_action_blocked((pe_resource_t *) (iter->data), task, reason);
+    }
```

Comment:
It's a bit unsettling. clone_internal_constraints and pcmk__bundle_internal_constraints() use pcmk__order_starts() with pe_order_runnable_left to order the collective pseudo-start before each real start. In those cases, pe_order_runnable_left ensures that if the pseudo-action can't run, the real ones can't either. But if you look at order_instance_promotion(), pseudo-promote before real promote is just optional.

It would probably be better to change the ordering flags wherever needed, but there are a ton of collective vs instance orderings that would have to be gone through carefully, and there would likely be unexpected consequences, so I'm not comfortable doing that in the rc phase. This PR is limited enough that I'm confident in the consequences.

But at some point that should be done, and then this could potentially be dropped.